v1/src/main/java/com/google/cloud/teleport/bigtable/CassandraColumnSchema.java

```diff
@@ -0,0 +1,139 @@
+ * command and is expected to be a string.
+ *
+ * <p>cqlsh -e "select json * from system_schema.columns where keyspace_name='$CASSANDRA_KEYSPACE'
+ * and table_name='$CASSANDRA_TABLE'"
+ *
+ * <p>This class expects only one table's columns as input.
+ *
+ * <p>This class is currently used to generate Cassandra writeTime queries.
+ */
+class CassandraColumnSchema {
+
+  private static final Logger LOG = LoggerFactory.getLogger(CassandraColumnSchema.class);
+  protected static final String KEYSPACE_NAME = "keyspace_name";
+  protected static final String TABLE_NAME = "table_name";
+  protected static final String POSITION = "position";
+
+  private String keyspace = "";
+  private String table = "";
+  private List<String> primaryKeyColumns;
+  private List<String> regularColumns;
+
+  /**
+   * Generate Cassandra column schema from a parsed schema string for use in creating a WriteTime
+   * query. The schema string format is expected to be the output of the following cqlsh command:
+   *
+   * <p>cqlsh -e "select json * from system_schema.columns where keyspace_name='$CASSANDRA_KEYSPACE'
+   * and table_name='$CASSANDRA_TABLE'"
+   *
+   * <p>This class expects only one table's columns as input.
+   *
+   * @param schemaString The parsed Cassandra columns schema string for one table.
+   */
+  CassandraColumnSchema(String schemaString) {
+    primaryKeyColumns = new ArrayList<>();
+    regularColumns = new ArrayList<>();
+
+    for (String line : schemaString.split("\\r?\\n")) {
+      // The schema file may not be a strict JSON. Skip lines that are not JSON rows.
+      if (!(line.contains(KEYSPACE_NAME)
+          && line.contains(TABLE_NAME)
+          && line.contains(COLUMN_NAME_COLUMN))) {
+        continue;
+      }
+      LOG.debug("Parsing JSON line " + line);
+
+      // Parse and set keyspace and table names.
+      JsonObject row = JsonParser.parseString(line).getAsJsonObject();
+      String rowKeyspace = row.get(KEYSPACE_NAME).getAsString();
+      String rowTable = row.get(TABLE_NAME).getAsString();
+      if (keyspace.isEmpty()) {
+        keyspace = rowKeyspace;
+      } else if (!keyspace.equals(rowKeyspace)) {
+        throw new IllegalArgumentException("Cassandra schema contains more than 1 keyspace_name");
+      }
+      if (table.isEmpty()) {
+        table = rowTable;
+      } else if (!table.equals(rowTable)) {
+        throw new IllegalArgumentException("Cassandra schema contains more than 1 table_name.");
+      }
+
+      // Parse primary and regular columns.
+      if (!row.has(POSITION) || !row.has(COLUMN_NAME_COLUMN)) {
+        throw new IllegalArgumentException("Cassandra schema row has no position or column_name.");
+      }
+      String rowColumnName = row.get(COLUMN_NAME_COLUMN).getAsString();
+      if (row.get(POSITION).getAsInt() > -1) {
+        primaryKeyColumns.add(rowColumnName);
+      } else {
+        regularColumns.add(rowColumnName);
```

Comment:
so regular column names have negative position values, and only the primary key have non negative position values?

Reply:
Correct - documentation on this is non-existent but the code points to this: https://github.com/apache/cassandra/blob/trunk/src/java/org/apache/cassandra/schema/ColumnMetadata.java#L92
This has been the case at least for the last 9 years (at least by blame history)